src/js/media-controller.js

```diff
@@ -62,57 +61,57 @@ class MediaController extends MediaContainer {
-      // some of these environment state values, we may want to model this
-      // a bit better (CJP).
-      volumeSupportPromise.then(() => {
-        this.propagateMediaState(
-          MediaUIAttributes.MEDIA_VOLUME_UNAVAILABLE,
-          MediaUIStates.MEDIA_VOLUME_UNAVAILABLE.get(this)
-        );
-      });
-    }
-
     // Track externally associated control elements
     this.mediaStateReceivers = [];
     this.associatedElementSubscriptions = new Map();
     this.associateElement(this);
 
-    // Apply ui event listeners to self
-    // Should apply to any UI container, not just controller
-    Object.keys(MediaUIRequestHandlers).forEach((key) => {
-      const handlerName = `_handle${constToCamel(key, true)}`;
-
-      // TODO: Move to map, not obj root property
-      this[handlerName] = (e) => {
-        // Stop media UI events from continuing to bubble
-        e.stopPropagation();
-
-        if (!this.media) {
-          console.warn('MediaController: No media available.');
-          return;
-        }
-
-        MediaUIRequestHandlers[key](this.media, e, this);
-      };
-      this.addEventListener(MediaUIEvents[key], this[handlerName]);
+    this._mediaStore = createMediaStore({
+      media: this.media,
+      // @ts-ignore
+      rootNode: this.getRootNode(),
+      fullscreenElement: this.fullscreenElement,
+      // NOTE: Names are hard. What should these be called (both the "bucket" currently called config and the props)
+      config: {
+        subtitles: this.hasAttribute('defaultsubtitles'),
+        mediaStreamType: this.getAttribute('defaultstreamtype') ?? undefined,
+        liveEdgeOffset: this.hasAttribute('liveedgeoffset') ? +this.getAttribute('liveedgeoffset') : undefined,
+        // NOTE: This wasn't updated if it was changed later. Should it be? (CJP)
+        noVolumePref: this.hasAttribute('novolumepref'),
+      }
     });
-
-    // Build event listeners for media states
-    this._mediaStatePropagators = {};
-    Object.keys(MediaUIStates).forEach((key) => {
-      this._mediaStatePropagators[key] = e => {
-        this.propagateMediaState(MediaUIProps[key], MediaUIStates[key].get(this, e));
-      };
+    Object.values(MediaUIEvents).forEach(eventName => {
+      this.addEventListener(eventName, e => {
+        // Pass events along to the media store.
+        // @ts-ignore
+        this._mediaStore.dispatch(e);
+      });
     });
-
+    const simpleStateCapture = () => {
+      let prevState = {};
+      return (nextState) => {
+        Object.entries(nextState).forEach(([stateName, stateValue]) => {
+          if (prevState[stateName] === stateValue) return;
+          // For now, we can just toLowerCase() from (camelCase) propName to (lowercase/"smushedcase") attrname
+          const attrName = stateName.toLowerCase();
+          this.propagateMediaState(attrName, stateValue);
+          // TODO: I don't think we want these events to bubble? Video element states don't. (heff)
```

Comment:
If we're gonna version bump I think we should change this with it. User "request" events I can see bubbling, but bubbling component state changes like this doesn't feel standard.